### PR TITLE
Enable Error Prone check: AnnotateFormatMethod

### PR DESCRIPTION
**Pull request by ksobolew**
## Description

The idea of it is simple: when it sees a pair of a format string and format arguments passes to a format method, and these two are parameters of a method, it asks the method to be annotated as a format methods as well. This works in tandem with `FormatStringAnnotation` (see also: #9512) as it allows it to detect more cases of malformed format strings and misused format methods (of which it found plenty in `trino-main` and `trino-spi`; most of them are missed opportunities to use the format method, but there are a couple of actual bad format strings).

Requires: #15051, ~~#15052~~

## Non-technical explanation

Better compile-time checks and better code quality.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ksobolew** [comment]
(I suppose I could split it into several smaller commits)

File: core/trino-main/src/main/java/io/trino/execution/CommentTask.java [hunk 1]
```diff
@@ -125,7 +125,7 @@ private void commentOnView(Comment statement, Session session)
             else if (metadata.getTableHandle(session, viewName).isPresent()) {
                 exceptionMessage += ", but a table with that name exists. Did you mean COMMENT ON TABLE " + viewName + " IS ...?";
             }
-            throw semanticException(TABLE_NOT_FOUND, statement, exceptionMessage);
+            throw semanticException(TABLE_NOT_FOUND, statement, "%s", exceptionMessage);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/execution/CommentTask.java:128]: Let's inline `exceptionMessage` so then you can do formatting at place

File: core/trino-main/src/main/java/io/trino/execution/CreateTableTask.java [hunk 2]
```diff
@@ -205,7 +205,7 @@ else if (element instanceof LikeClause) {
                     if (!originalLikeTableName.equals(likeTableName)) {
                         message += format(". LIKE table '%s' redirected to '%s'.", originalLikeTableName, likeTableName);
                     }
-                    throw semanticException(NOT_SUPPORTED, statement, message);
+                    throw semanticException(NOT_SUPPORTED, statement, "%s", message);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/execution/CreateTableTask.java:208]: same

File: testing/trino-testing-services/src/main/java/io/trino/testng/services/Listeners.java [hunk 3]
```diff
@@ -24,8 +25,9 @@ private Listeners() {}
     /**
      * Print error to standard error and exit JVM.
      *
-     * @apiNote A TestNG listener cannot throw an exception, as this are not currently properly handlded by TestNG.
+     * @apiNote A TestNG listener cannot throw an exception, as this are not currently properly handled by TestNG.
```
Comment (kokosing) [inline testing/trino-testing-services/src/main/java/io/trino/testng/services/Listeners.java:28]: separate commit

**Comment by kokosing** [comment]
> I suppose I could split it into several smaller commits

👍

**Comment by ksobolew** [comment]
There you go: 14 commits :)

File: core/trino-main/src/main/java/io/trino/execution/CommentTask.java (hunk 1 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/CommentTask.java]: Done, but only partially. The additional information could also be a format string, but it would require using different `throw`s in each branch, or some other tricks.

File: core/trino-spi/pom.xml [hunk 4]
```diff
@@ -35,6 +35,11 @@
             <optional>true</optional>
         </dependency>
 
+        <dependency>
```
Comment (kokosing) [inline core/trino-spi/pom.xml:38]: Adding dependency to SPI is serious. It needs to go in separate PR.

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java [hunk 5]
```diff
@@ -110,7 +110,7 @@ public ListenableFuture<Void> execute(
             propertyMetadata.decode(objectValue);
         }
         catch (RuntimeException e) {
-            throw semanticException(INVALID_SESSION_PROPERTY, statement, e.getMessage());
+            throw semanticException(INVALID_SESSION_PROPERTY, statement, "%s", e.getMessage());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: I don't like it. It looks like hacky suppress, I think the previous version is clearer and it should be allowed.

File: core/trino-spi/pom.xml (hunk 4 above)
Comment (ksobolew) [reply core/trino-spi/pom.xml:38]: It is, kind of. It can't be a separate commit entirely because then all the enforcers etc. will complain that the dependency is not used.

**Review by kokosing — COMMENTED** [review]
LGTM for the below commits and the first commit. If you could extract them as separate PR I can merge it.

I left two comments that are required to move forward. I haven't reviewed entire PR.

image

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 5 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: But it's not clear to the static analyzers, which want to make sure that the message doesn't malformed format strings, like `%` signs which will error out in `String#format`. It's also not clear to me, because by looking at it I have no idea that it's a format method and I should be wary of possible errors.

There *is* a possibility to add an overload without the format parameters; it won't use `String#format` and it won't be annotated as a format method. But you lose the ability to detect mis-uses of the format method, like:

```java
            throw semanticException(INVALID_SESSION_PROPERTY, statement, "Nested error: " + e.getMessage());
```

which should be:

```java
            throw semanticException(INVALID_SESSION_PROPERTY, statement, "Nested error: %s", e.getMessage());
```

instead.

File: core/trino-spi/pom.xml (hunk 4 above)
Comment (ksobolew) [reply core/trino-spi/pom.xml:38]: Oh, sorry, you said "PR", I read "commit" :)

**Comment by ksobolew** [comment]
> LGTM for the below commits and the first commit. If you could extract them as separate PR I can merge it.

Extracted as #15051

File: core/trino-spi/pom.xml (hunk 4 above)
Comment (ksobolew) [reply core/trino-spi/pom.xml:38]: Extracted as #15052

Comment (kokosing) [reply core/trino-spi/pom.xml:38]: As I understand, the extracted PR blocks this PR, right?

Comment (ksobolew) [reply core/trino-spi/pom.xml:38]: In a sense, yes. You could still merge this one as it includes the other one and the result will be the same :)

**Comment by ksobolew** [comment]
Flaky hit: #14487

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 5 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: You cannot do this. It is an user facing exception message, so you want the nice and actionable message. This would be a regression.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 6]
```diff
@@ -1018,15 +1018,15 @@ else if (physicalOffset < 0) {
             return builder.toString();
         }
 
-        private String formatRowsPerMatch(RowsPerMatch rowsPerMatch)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1021]: I liked the previous more as it was more generic

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 7]
```diff
@@ -1029,14 +1029,14 @@ private void formatRowsPerMatch(RowsPerMatch rowsPerMatch, NodeRepresentation no
             }
         }
 
-        private String formatSkipTo(Position position, Optional<IrLabel> label)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1032]: same

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 5 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: But it doesn't change the message, just makes sure that it won't be misinterpreted as a format string...

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java (hunk 6 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1021]: How is it more generic? It seems perfectly equivalent to me...

Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1021]: The alternative is to use the `nodeOutput.appendDetails("%s", formatRowsPerMatch(...))` pattern :)

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 5 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: Maybe `"Invalid session property: %s"`? Or maybe using `semanticException` here is a mistake and we should just `throw new TrinoException`?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java [hunk 8]
```diff
@@ -792,7 +791,7 @@ private void verifyNoOrderByReferencesToOutputColumns(Node node, StandardErrorCo
         getReferencesToScope(node, analysis, orderByScope.get())
                 .findFirst()
                 .ifPresent(expression -> {
-                    throw semanticException(errorCode, expression, errorString);
+                    throw semanticException(errorCode, expression, "%s", errorString);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java:794]: same

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 9]
```diff
@@ -3163,7 +3166,9 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Nod
                 return superType;
             }
 
-            throw semanticException(TYPE_MISMATCH, node, message, firstType, secondType);
+            @SuppressWarnings("FormatStringAnnotation") // Error Prone wants the types of format arguments to be the same as where @FormatString is declared, but we need them to be different
+            TrinoException exception = semanticException(TYPE_MISMATCH, node, message, firstType, secondType);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:3170]: new TrinoException?

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java (hunk 6 above)
Comment (kokosing) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1094]: It was more generic as you could anything with the returned value. Now the value is always consumed the same way.

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 5 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: That's a good point. Changed that to mention the property name as well.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java (hunk 8 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/AggregationAnalyzer.java:794]: In this case I annotated the enclosing method as a `@FormatMethod` instead. It's more generic this way ;)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 9 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:3170]: I don't know, I think it's valid, just hits some limitations of Java type system (or Error Prone's interpretation of it)

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java (hunk 6 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:1094]: OK, I'll drop the two commits which refactor these methods and I'll "paste" the return value into a simple format string instead.

**Comment by ksobolew** [comment]
I also added (at the beginning) a small-ish commit which changes some logger calls to use formatted messages. I though it was relevant (it also makes it easier for me to work on the main branch, because in my setup Error Prone complains about these).

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java [hunk 10]
```diff
@@ -571,7 +571,7 @@ private <T> T executeWithSession(SessionCallable<T> sessionCallable)
                     throw e;
                 }
                 long delay = Math.min(schedule.nextDelay().toMillis(), timeLeft);
-                log.warn(e.getMessage());
```
Comment (findepi) [inline plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:574]: does `log.warn(String)` do any interpolation?

File: core/trino-main/src/main/java/io/trino/execution/SetColumnTypeTask.java [hunk 11]
```diff
@@ -78,15 +77,18 @@ public ListenableFuture<Void> execute(
...
             else if (metadata.getView(session, qualifiedObjectName).isPresent()) {
-                exceptionMessage += ", but a view with that name exists.";
+                additionalInformation = ", but a view with that name exists.";
+            }
+            else {
+                additionalInformation = "";
             }
             if (!statement.isTableExists()) {
-                throw semanticException(TABLE_NOT_FOUND, statement, exceptionMessage);
+                throw semanticException(TABLE_NOT_FOUND, statement, "Table '%s' does not exist%s", qualifiedObjectName, additionalInformation);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/SetColumnTypeTask.java:91]: just do
```
throw semanticException(TABLE_NOT_FOUND, statement, "%s", exceptionMessage);
```
(reverting other changes)

i think it will be more readable

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java [hunk 12]
```diff
@@ -110,7 +110,7 @@ public ListenableFuture<Void> execute(
             propertyMetadata.decode(objectValue);
         }
         catch (RuntimeException e) {
-            throw semanticException(INVALID_SESSION_PROPERTY, statement, e.getMessage());
+            throw semanticException(INVALID_SESSION_PROPERTY, statement, "Invalid session property value '%s': %s", propertyName.toString(), e.getMessage());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: SET SESSION is invoked for a particular session property.
we don't need to provide the property name in the exception message, do we?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 13]
```diff
@@ -3137,7 +3139,8 @@ private void coerceType(StackableAstVisitorContext<Context> context, Expression
             coerceType(expression, actualType, expectedType, message);
         }
 
-        private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Node node, String message, Expression first, Expression second)
+        @FormatMethod
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:3142]: Technically, doesn't belong to `Annotate semanticException factories as format methods` commit

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java (hunk 9 above)
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:3170]: instead of suppression, just do this
```
throw semanticException(TYPE_MISMATCH, node, "%s", format(message, firstType, secondType));
```

**Review by findepi — COMMENTED** [review]
I stopped reviewing before "Make some conditional uses of NodeRepresentation#appendDetails explicit"

I know i will have to re-review everything. Please split the PR into smaller, so that we don;t have to re-review same code multiple times.

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java (hunk 10 above)
Comment (ksobolew) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:574]: > does `log.warn(String)` do any interpolation?

Not at this moment. I made a change upstream to make it easier to detect places like this, but it's a semantics change.

**Comment by ksobolew** [comment]
> I know i will have to re-review everything. Please split the PR into smaller, so that we don;t have to re-review same code multiple times.

Sure. First two: #15962 and #15963

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java (hunk 10 above)
Comment (findepi) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:574]: we should not do a semantics change.
what's the problem with `log.warn(String)` not doing any interpolation?

File: core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java (hunk 12 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/SetSessionTask.java:113]: awesome. can this be a separate change, or even separate PR?
this change is orthognal to error-prone

**Comment by ksobolew** [comment]
Next one: #15964

File: plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java (hunk 10 above)
Comment (ksobolew) [reply plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraSession.java:574]: > what's the problem with `log.warn(String)` not doing any interpolation?

It's arguably inconsistent with the overload which does take parameters. If this overload didn't exist, or if it would be marked as a format method, it would behave the same way the `String.format("abcde")` behaves, which does do interpolation (the only interpolation it can do is `"%%"`, though). But this is very arguable and people's expectations differ. I would expect this to be consistent across all invocations, but other people don't like it.

But the primary reason I would like `log.warn(String)` to be a format method is to allow Error Prone detect cases where the log message is constructed using concatenation. `log.warn(e.getMessage());` is a casualty of this, unfortunately :)
